packages/flutter/lib/src/painting/placeholder_span.dart

```diff
@@ -0,0 +1,74 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui' as ui show ParagraphBuilder, PlaceholderAlignment;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/gestures.dart';
+import 'package:flutter/services.dart';
+
+import 'basic_types.dart';
+import 'inline_span.dart';
+import 'text_style.dart';
+import 'text_painter.dart';
+import 'text_span.dart';
+
+/// An immutable placeholder that is embedded inline within text.
+///
+/// [PlaceholderSpan] represents an abstract generic placeholder. [WidgetSpan] should be
+/// used instead to specify the contents of the placeholder. A [PlaceholderSpan]
+/// by itself does not contain useful information to change a TextSpan.
+///
+/// [PlaceholderSpan]s must be leaf nodes in the [TextSpan] tree. [PlaceholderSpan]s
```

Comment:
saying "must be" makes it sound like it's possible to do anything but this. Does it mean that it would be an error for a subclass to define a way to have children? why would that be so bad?

Reply:
This probably belongs in the WidgetSpan docs instead of here. I should also clarify in the WidgetSpan docs that the no-children only applies to children of the type InlineSpan.

Reply:
I could also just allow everything to have children and handle it the same way TextSpan does (children get the parent TextStyle when not explicitly specified)

Reply:
Assuming we implement this well (i.e. not using `is` or `as` anywhere), you should be able to create an entirely new subclass of `InlineSpan` that does everything `TextSpan` does, including having children, doing accessibility, etc, and it should just work. That means that `PlaceholderSpan`s could in fact have children in the `InlineSpan` sense and there's no reason it wouldn't work.

It's not really a matter of "allowing" or "disallowing". I think it's clear that _our_ `WidgetSpan` class shouldn't have `TextSpan` children -- what would that even mean... after all, the child of the `WidgetSpan` is a `Widget. But if `TextSpan` can support children, then why wouldn't other subclasses of `InlineSpan` be unable to have them.

Reply:
So I think the motivation of the `children` API is to have multiple spans share the same TextStyle. Since TextStyles do have a purpose when applied to WidgetSpans (eg, decoration control, backgrounds/foregrounds), it WidgetSpans can logically have children that simply share the parent's TextStyle. The child WidgetSpans would behave like the child TextSpans, which render inline after/downstream from the parent's contents.